sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -267,6 +266,23 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
 
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
       provider.getTable(dsOptions) match {
+        // TODO: for backward compatibility reasons, the builtin file source needs to support all
+        // the save modes, which violates the semantic of `TableProvider`. Here we special-case
+        // file source and pass the save mode to file source directly. This hack can be removed
+        // after we figure out a general interface for path-based data sources.
+        case table: FileTable =>
```

Comment:
Why is this hack necessary? Why not put off v2 support for path-based tables?

Reply:
Because we want to migrate file source to v2, to validate the API of `Table`, `ScanBuilder`, `WriteBuilder`, etc.

This hack is just to work around the issue that we do not have a proper entry API for path-based data source, which I believe we will have later on.

I think this is not a bad idea. We can unblock the file source migration, and we can keep the DS v2 clean (`FileTable` is an internal class). Besides that this hack will be removed once we have path-based API in DS v2.

Reply:
File sources do not use v2 by default, so this is not a necessary change for this commit. I think it should be in a separate PR. We can discuss whether it is a good idea to add this hack in the next DSv2 sync.

Please remove it from this PR.

Reply:
I'm adding this to my agenda for the sync-up.

Reply:
I believe that we agreed in the sync-up to remove this hack. Is there a reason why it is still included?

Reply:
I explained in the comment: https://github.com/apache/spark/pull/24233#issuecomment-492628836

Removing the hack breaks several tests and I'd like to do it in another PR. Since `SupportsSaveMode` is a hack, which means we already have a hack for file source v2 in `DataFrameWriter.save`. To keep the PR small, I think it's better to still keep the hack(it's not creating a new hack) and open a new PR to remove it and update the tests.